active/XamlServiceProvider/XamlServiceProviderApi.md

```diff
@@ -0,0 +1,272 @@
+
+# Background
+
+_This spec adds a serviceProvider to the existing [MarkupExtension](https://docs.microsoft.com/uwp/api/Windows.UI.Xaml.Markup.MarkupExtension) class. ([Proposal](https://github.com/microsoft/microsoft-ui-xaml/issues/741))_
+
+Xaml markup has a [markup extension](https://docs.microsoft.com/en-us/dotnet/desktop-wpf/xaml-services/markup-extensions-overview) feature which lets you define a builder which is called by the Xaml loader.
+
+When you write a custom markup extension, you override the `ProvideValue` method to return the built value (see the next section for an example). The ProvideValue method is different in WinUI than it is in WPF, because WPF passes an extra [IServiceProvider](http://msdn.microsoft.com/library/System.IServiceProvider) parameter:
+
+WinUI:
+
+```cs
+protected override object ProvideValue()
+{
+    return ...
+}
+```
+
+WPF:
+
+```cs
+protected override object ProvideValue(IServiceProvider serviceProvider)
+{
+    return ...
+}
+```
+
+The service provider can be used, for example, to discover the property being assigned to.
+
+This spec is adding a new version of ProvideValue to WinUI that passes a similar service provider. All the APIs match WPF, except where WPF uses the System.IServiceProvider, WinUI will use Microsoft.UI.Xaml.Markup.IXamlServiceProvider.
+
+> To do: Add an Issue to cs/winrt to project IServiceProvider to .Net5 as [System.IServiceProvider](http://msdn.microsoft.com/library/System.IServiceProvider)?

```

Comment:
Also, should this be in the Xaml namespace or someplace more general?

Reply:
What namespace did we put our versions of INPC and INCC under?

Reply:
Xaml. But for example we moved [Color](https://docs.microsoft.com/uwp/api/Windows.UI.Color) out of the Xaml namespace, since it's useful outside that one domain. There's nothing specific to Xaml about the service provider pattern.

Reply:
I think what I was trying to ask is: how does CsWinrt project INPC for a non-WinUI project? Is that project still required to include WinUI so that it knows about Microsoft.UI.Xaml.Data.INotifyPropertyChanged? Or is that a nonsensical question?

Reply:
That's a sensical question; I don't think INPC should have been put into the Xaml namespace.

Reply:
Current plan is to leave this as is in this spec (in the Xaml.Data namespace).